sympy/polys/matrices/tests/test_rref.py

```diff
@@ -25,24 +34,42 @@
         'zz_3',
         DM([[1, 2],
             [3, 4]], ZZ),
-        DM([[-2,  0],
-            [ 0, -2]], ZZ),
-        ZZ(-2),
+        DM([[1, 0],
+            [0, 1]], ZZ),
+        ZZ(1),
```

Comment:
With the addition of sparse fraction-free RREF and after further investigation I now realise that the fraction-free methods do not produce results with a reliable denominator. For example just reordering the rows can change the denominator. These tests are changed to test only the cancelled form (after using `DomainMatrix.cancel_denom`). Probably the functions that return results with denominator like `solve_den` and `inv_den` should use `cancel_denom` by default so that they can return canonical results. At least canonical results are needed to use these as lower-level routines from user-facing front-end functions like `linsolve` and `Matrix.inv`.

Reply:
>  just reordering the rows can change the denominator

Can the rows be put in canonical order at the outset?